### PR TITLE
QuickAI & nntrainer Integration

### PR DESCRIPTION
**Pull request by jayden0701**
This PR brings nntrainer to a state where it can power **QuickAI** end-to-end. The changes span the CausalLM pipeline, QNN backend integration, chat template handling, and core engine fixes.

### Highlights

**QuickAI Readiness**
This version includes all the building blocks needed to run QuickAI: a multimodal embedding interface, embedding sidecar with LUT, logits processor hooks, an explicit cancel API, and a vendored QNN backend — making the full inference pipeline self-contained.

### New Features

- **Multimodal embedding interface** — Model-agnostic multimodal embedding interface in the transformer layer, enabling vision+language inputs.
- **Embedding sidecar with LUT** — Adds an embedding sidecar layer with lookup-table support and config wiring for flexible token/vision embedding.
- **Generic logits processor hook** — Pluggable logits processor hook in CausalLM, allowing custom post-processing (e.g., repetition penalty, top-k filtering).
- **Explicit cancel API** — Adds an explicit cancellation API to CausalLM for aborting in-flight generation.
- **QNN rpcmem allocator** — Tensor-only rpcmem allocator for the QNN activation pool, improving memory management on device.

### QNN Backend

- **Vendored QNN backend** — Hosts the QNN backend directly inside nntrainer, removing the external dependency.
- **Drop ml-api-inference link dep** — `libqnn_context.so` no longer depends on ml-api-inference, simplifying the build graph.
- **Android build guard** — Replaces `exit(-1)` with `throw` and adds Android build guards for robustness.

### Chat Template

- **Vendored minja** — Replaces the minja git submodule with vendored header files for reproducible builds.
- **Capitalize filter** — Bumps minja to add a `capitalize` template filter.

### Bug Fixes

- **Engine singleton** — Makes `Engine::Global()` a true singleton and ensures registries are idempotent, preventing duplicate registration on re-initialization.
- **Multi-token prefill** — Generalizes `incremental_forwarding` to correctly handle multi-token prefill sequences.
- **AppContext tests** — Aligns clashing-key tests with the new no-op re-registration behavior.

### Refactoring

- **UTF-8 streaming guard** — Extracts the UTF-8 streaming guard into a shared header (`utf8_stream_util.h`) for reuse across components.

---

### CODE REVIEW

File: Applications/CausalLM/api/callback_streamer.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @author  Joonseok Oh <jrock.oh@samsung.com>
+ * @bug     No known bugs except for NYI items
+ */
+
+#include "callback_streamer.h"
+
+extern "C" {
+
+static int callback_streamer_put(BaseStreamer *self, const char *decoded_utf8) {
+  auto *streamer = reinterpret_cast<CallbackStreuamer *>(self);
```
[the remark this answers is not in the document]
Comment (jayden0701) [reply Applications/CausalLM/api/callback_streamer.cpp]: Will fix in future quick api -> causallm api patch

[the remark this answers is not in the document]
Comment (jaemini-shin) [reply Applications/CausalLM/api/callback_streamer.cpp]: Please fix this in this PR. It will affect every other PR's CI.

[the remark this answers is not in the document]
Comment (jayden0701) [reply Applications/CausalLM/api/callback_streamer.cpp]: Oh CI fails...I'll fix it

**Review by dlwlzzero — COMMENTED** [review]
Thanks for your work 👍 However, it seems a conflict has occurred; could you resolve it?

**Comment by jayden0701** [comment]
> Thanks for your work 👍 However, it seems a conflict has occurred; could you resolve it?

Oh #4016 was merged after this PR.... I will rebase soon

**Comment by jijoongmoon** [comment]
Since we have update_qnn_version.py‎, we could remove the QNN headers. Could you consider it?

**Review by jijoongmoon — APPROVED** [review]
LGTM. Excpet minor

**Review by jaemini-shin — APPROVED** [review]
LGTM, but please make PR as small as possible next time

**Comment by jayden0701** [comment]
> LGTM, but please make PR as small as possible next time

Since This PR is a migration from a huge PR #3963 , It was my best :(

**Comment by jayden0701** [comment]
> Since we have update_qnn_version.py‎, we could remove the QNN headers. Could you consider it?

I personally thought that if we don't keep the QNN version to a specific one, we could face some bugs.
So maybe we should remove headers but keep the version fixed?

**Review by Jungwon-Lee — APPROVED** [review]
LGTM! Thank you for your effort!

**Review by baek2sm — APPROVED** [review]
LGTM

**Comment by jijoongmoon** [comment]
Could you check the cpp-linter CI?
